feat: support for hosted token worker

DJMcK opened the pull request:
<!-- By submitting a PR to this repository, you agree to the terms within the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md). Please see the [contributing guidelines](https://github.com/auth0/.github/blob/master/CONTRIBUTING.md) for how to create and submit a high-quality PR for this repo. -->

### Changes

When the SDK is used in combination with a strict Content-Security-Policy (CSP), the policy must include `worker-src: blob:` which raises a concern of `unsafe-eval`. This change allows the SDK to be configured to load the worker code from a trusted URL, compliant with the CSP, and allows the user to mitigate the concern.

**Todo:**
- [x] Just waiting for https://github.com/auth0/auth0-spa-js/pull/1209 to land so that we can get e2e coverage of this 

### References

> As defined above, special URL schemes that refer to specific pieces of unique content, such as `data:`, `blob:` and `filesystem:` are excluded from matching a policy of * and must be explicitly listed. Policy authors should note that the content of such URLs is often derived from a response body or execution in a Document context, which may be unsafe. Especially for the [default-src](https://www.w3.org/TR/CSP2/#default_src) and [script-src](https://www.w3.org/TR/CSP2/#script_src) directives, policy authors should be aware that allowing `data:` URLs is equivalent to unsafe-inline and **allowing `blob:` or `filesystem:` URLs is equivalent to unsafe-eval**.

https://www.w3.org/TR/CSP2/#source-list-guid-matching

### Testing

<!--
Please describe how this can be tested by reviewers. Be specific about anything not tested and reasons why. If this library has unit and/or integration testing, tests should be added for new functionality and existing tests should complete without errors.
-->

- [X] This change adds unit test coverage
- [ ] This change adds integration test coverage
- [X] This change has been tested on the latest version of the platform/language

### Checklist

- [X] I have read the [Auth0 general contribution guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md)
- [X] I have read the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md)
- [X] All code quality tools/guidelines have been run/followed

frederikprijck reviewed (changes requested):
Thanks for this, this looks realy good.

The only thing that confuses me a bit is that it looks like we are now distributing a separate file for the worker. How is the user supposed to use this? I assume they need to ensure they take the file from their node_modules, and copy it as-is to their web-server without bundling? Can we add some clarification to the PR? Or maybe add something in `Examples.md`?

It also looks like whoever opts to use the worker by URL, will have two copies of the worker (one in our bundle, one in the separate file), is that accurate and expected?

DJMcK commented:
Hey @frederikprijck 👋 

> I assume they need to ensure they take the file from their node_modules, and copy it as-is to their web-server without bundling?

That's right, or leverage our CDN distribution of the `...worker.production.js` file. 

There's an effort to explain that to the consumer via the [`src/global.ts`](https://github.com/auth0/auth0-spa-js/pull/1208/files#diff-0bec2401f35d4c80319a0256eadb80621ce9a826423909629bb0708b7758307d) docs. Tough one to articulate in a few sentences, I'll add a bit more in `Examples.md` too.

> It also looks like whoever opts to use the worker by URL, will have two copies of the worker (one in our bundle, one in the separate file), is that accurate and expected?

Yeah, that's right. You'd have the bundled `auth0-spa-js.production.js` which includes the integrated worker code and, if you opt to not use the eval approach, then you'd load the separate `auth0-spa-js.worker.production.js`. 

There's definitely a tradeoff being made between a bit of performance for security when you opt-in to the latter.

frederikprijck commented on src/global.ts:
```diff
@@ -259,6 +259,16 @@ export interface Auth0ClientOptions extends BaseLoginOptions {
+   * at `<module_path>/dist/auth0-spa-js.worker.production.js`, from the same host as your application or using the Auth0 CDN 
+   * `https://cdn.auth0.com/js/auth0-spa-js/<version>/auth0-spa-js.worker.production.js`.
+   * 
+   * **Note**: The worker is only used when `useRefreshTokens: true` and no custom `cache` is provided.
```
I think this is not correct. Based on the [code](https://github.com/auth0/auth0-spa-js/blob/main/src/Auth0Client.ts#L232), it should be only when useRefreshTokens is true, and the cache is in memory (so not local-storage or custom).

DJMcK replied:
Thanks for catching that! Just made an update to it in [03fb78d](https://github.com/auth0/auth0-spa-js/pull/1208/commits/03fb78d6aaefdf2c4b4f7322c5acf237a477dbe7).

DJMcK commented:
@frederikprijck I took a pass at an FAQ in https://github.com/auth0/auth0-spa-js/pull/1208/commits/1691d42098cd9e2d278b9513cf88c85831050214. Let me know your thoughts on that draft.

frederikprijck reviewed (changes requested):
Looks good, but lets add the e2e tests we discussed on slack.

DJMcK commented:
- Rebased to get the head on https://github.com/auth0/auth0-spa-js/pull/1209 (integration tests fix)
- Added the new integration test for this feature https://github.com/auth0/auth0-spa-js/pull/1208/commits/9044c636fcf41ac36d343ca0c90f676becf1d249
